src/util/util_unix.go

```diff
@@ -5,29 +5,67 @@ package util
 import (
 	"os"
 	"os/exec"
+	"strings"
 	"syscall"
 
 	"golang.org/x/sys/unix"
 )
 
-// ExecCommand executes the given command with $SHELL
-func ExecCommand(command string, setpgid bool) *exec.Cmd {
+type Executor struct {
+	shell   string
+	args    []string
+	escaper *strings.Replacer
+}
+
+func NewExecutor(withShell string) *Executor {
 	shell := os.Getenv("SHELL")
-	if len(shell) == 0 {
-		shell = "sh"
+	args := strings.Fields(withShell)
+	if len(args) > 0 {
+		shell = args[0]
+		args = args[1:]
+	} else {
+		if len(shell) == 0 {
+			shell = "sh"
+		}
+		args = []string{"-c"}
 	}
-	return ExecCommandWith(shell, command, setpgid)
+
+	var escaper *strings.Replacer
+	tokens := strings.Split(shell, "/")
+	if tokens[len(tokens)-1] == "fish" {
+		// https://fishshell.com/docs/current/language.html#quotes
+		// > The only meaningful escape sequences in single quotes are \', which
+		// > escapes a single quote and \\, which escapes the backslash symbol.
+		escaper = strings.NewReplacer("\\", "\\\\", "'", "\\'")
+	} else {
+		escaper = strings.NewReplacer("'", "'\\''")
+	}
+	return &Executor{shell, args, escaper}
 }
 
-// ExecCommandWith executes the given command with the specified shell
-func ExecCommandWith(shell string, command string, setpgid bool) *exec.Cmd {
-	cmd := exec.Command(shell, "-c", command)
+// ExecCommand executes the given command with $SHELL
+func (x *Executor) ExecCommand(command string, setpgid bool) *exec.Cmd {
+	cmd := exec.Command(x.shell, append(x.args, command)...)
 	if setpgid {
 		cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
 	}
 	return cmd
 }
 
+func (x *Executor) QuoteEntry(entry string) string {
+	return "'" + x.escaper.Replace(entry) + "'"
+}
+
+func (x *Executor) Become(stdin *os.File, environ []string, command string) {
+	SetStdin(stdin)
+	shellPath, err := exec.LookPath(x.shell)
```

Comment:
This works as long as the user provides the full path to their binary.

```bash
fzf --with-shell '/usr/bin/ruby -e' --bind 'enter:become:puts "Hello world"'
```
However, if the user only provides the name, then the execution stalls for the `become` part.

```bash
fzf --with-shell 'ruby -e' --bind 'enter:become:puts "Hello world"'
```

How about this ?

```suggestion
	if err != nil {
		return
	}
```

Reply:
or better use `os.Exit(...)` ?

```bash
fzf --with-shell 'idontexist -e' --bind 'enter:become:puts "Hello world"'
```

Reply:
Thanks, it was an apparent bug. I'll make it exit with 127.

```sh
$ sh -c 'nope'
sh: nope: command not found
$ echo $?
127
```

Reply:
Maybe we should print such an error message as well?

Reply:
```
$ fzf --with-shell 'ruby2 -e' --bind 'start:become:puts "Hello world"'
fzf (become): exec: "ruby2": executable file not found in $PATH
```

Reply:
Thanks, the error message is useful.

In both `zsh`[^1] and `bash`[^2], there is a distinction between exit code `127` (command not found) and `126` (command found but not executable). It is unclear if you wish to apply this distinction here.

```bash
sh -c "$HOME"
# sh: /Users/paria: is a directory
echo $?
# 126
```

If not, simply returning error code `2` would suffice without the need to modify the man page for the exit status. 

https://github.com/junegunn/fzf/blob/bacb7d3ba301094bd7e108997193860712ec5f32/man/man1/fzf.1#L938-L945

[^1]: [zsh: 8 Command Execution](https://zsh.sourceforge.io/Doc/Release/Command-Execution.html#Command-Execution)
[^2]: [Bash Reference Manual - Exit Status](https://www.gnu.org/software/bash/manual/bash.html#Exit-Status)